app/controllers/samples_controller.rb

```diff
@@ -721,30 +722,78 @@ def samples_going_public
     render json: samples.to_json(include: [{ project: { only: [:id, :name] } }])
   end
 
+  # The json response here should be precached in PipelineRun.
   def report_info
-    expires_in 30.days
-    @pipeline_run = select_pipeline_run(@sample, params)
+    MetricUtil.put_metric_now("samples.cache.requested", 1)
 
+    pipeline_run = select_pipeline_run(@sample, params[:pipeline_version])
+    if pipeline_run.nil?
+      message = "Pipeline run not found for sample #{@sample.id}"
+      raise ActiveRecord::RecordNotFound, message
+    end
+    report_info_params = pipeline_run.report_info_params
+
+    # Set background_id to a viewable background
+    params[:background_id] = get_background_id(@sample)
+
+    cache_key = self.class.report_info_cache_key(
+      request.path,
+      params.permit(report_info_params.keys)
+    )
+
+    # This allows 304 Not Modified to be returned so that the client can use its
+    # local cache and avoid the large download.
+    httpdate = Time.at(report_info_params[:report_ts]).utc.httpdate
+    response.headers["Last-Modified"] = httpdate
+    # This is a custom header for testing and debugging
+    response.headers["X-IDseq-Cache"] = 'requested'
+
+    pipeline_run = select_pipeline_run(@sample, params[:pipeline_version])
+    json = Rails.cache.fetch(cache_key) do
+      MetricUtil.put_metric_now("samples.cache.miss", 1)
+      response.headers["X-IDseq-Cache"] = 'missed'
+      SamplesController.report_info_json(pipeline_run, params[:background_id])
```

Comment:
it looks a little odd to call controller methods statically, so report_info_json could also move to ReportHelper

Reply:
good point